splitgpg2/__init__.py

```diff
@@ -0,0 +1,704 @@
+
+    def __init__(self, reader: asyncio.StreamReader,
+                 writer: asyncio.StreamWriter, client_domain: str):
+
+        # configuration options:
+        self.verbose_notifications = False
+        self.timer_delay = self.default_timer_delay()
+        #: allow client to generate a new key
+        self.allow_keygen = False
+        #: signal those Futures when connection is terminated
+        self.notify_on_disconnect = set()
+        self.log_io_enable = False
+
+        self.client_reader = reader
+        self.client_writer = writer
+        self.client_domain = client_domain
+        self.commands = self.default_commands()
+        self.options = self.default_options()
+        self.hash_algos = self.default_hash_algos()
+
+        self.log = logging.getLogger('splitgpg2.Server')
+        self.agent_socket_path = None
+        self.agent_reader = None
+        self.agent_writer = None
+        self.untrusted_input_buffer = bytearray()
+
+        debug_log = os.environ.get('QUBES_SPLIT_GPG2_DEBUG_LOG', None)
+        if debug_log:
+            handler = logging.FileHandler(debug_log)
+            self.log.addHandler(handler)
+            self.log.setLevel(logging.DEBUG)
+            self.log_io_enable = True
+
+    async def run(self):
+        await self.connect_agent()
+        try:
+            while not self.client_reader.at_eof():
+                await self.handle_command()
+        finally:
+            # close connection to the real gpg agent too
+            self.agent_writer.close()
+
+    def log_io(self, prefix, untrusted_msg):
+        if not self.log_io_enable:
+            return
+        allowed = string.printable.\
+            replace('\n', '').\
+            replace('\r', '').\
+            replace('\f', '').\
+            replace('\v', '')
+        allowed = allowed.encode('ascii')
+        self.log.warning('%s: %s', prefix, ''.join(
+            chr(c) if c in allowed else '.'
+            for c in untrusted_msg.strip()))
+
+    async def connect_agent(self):
+        try:
+            subprocess.check_call(['gpgconf', '--launch', 'gpg-agent'])
+        except subprocess.CalledProcessError:
+            raise StartFailed
+
+        dirs = subprocess.check_output(['gpgconf', '--list-dirs'])
+        # search for agent-socket:/run/user/1000/gnupg/S.gpg-agent
+        agent_socket_path = [d.split(b':', 1)[1] for d in dirs.splitlines()
+                             if d.startswith(b'agent-socket:')][0]
+        self.agent_socket_path = agent_socket_path.decode()
+
+        self.agent_reader, self.agent_writer = await asyncio.open_unix_connection(
+                path=self.agent_socket_path)
+
+        if self.verbose_notifications:
+            self.notify('connected')
+
+        # wait for agent hello
+        await self.handle_agent_response()
+
+    def abort(self, reason):
+        self.log.error('%s; Aborting!', reason)
+        self.client_writer.close()
+        self.agent_writer.close()
```

Comment:
Don't you need to also close the reader?

To me an explicit exit (or similar) would be more obvious. If you want to keep it as-is please add a comment why closing is enough to abort.

Reply:
Closing writer is enough - [it closes unrelying socket fd](https://docs.python.org/3/library/asyncio-stream.html#asyncio.StreamWriter.close), there is no such thing as reader.close().
I'll add a comment.

Yes, I want to avoid exit(), to be allow serving multiple clients in the same process (avoid startup overhead).

Reply:
Added a comment in my branch. Also renamed the function.

> Closing writer is enough - it closes unrelying socket fd, there is no such thing as reader.close().

Oh, ok. If I understand the docs correctly reader and writer are always using the same socket. But why are they two different object? I don't like that the API of `__init__` allows passing arguments which violate our assumption. Can we check that we got the reader that matches the writer?

Reply:
> I don't like that the API of `__init__` allows passing arguments which violate our assumption.

Unfortunately in asyncio those are two objects for some (unknown to me) reason. Same is true when [handling listening socket](https://docs.python.org/3/library/asyncio-stream.html#asyncio.start_server) - you provide a callback that gets reader and writer.

> Can we check that we got the reader that matches the writer?

I don't see any official way for that. Unofficial is `reader._transport == writer.transport`.

Reply:
Do you prefer to leave it as is or to use the unofficial check? I'm fine either way.

Reply:
I'd leave it as is.